CelesteNet.Shared/DataTypes/DataPlayerFrame.cs

```diff
@@ -0,0 +1,270 @@
+
+        public int CurrentAnimationID;
+        public int CurrentAnimationFrame;
+
+        public Color[] HairColors = Dummy<Color>.EmptyArray;
+        public string HairTexture0 = string.Empty;
+        public bool HairSimulateMotion;
+
+        public Entity[] Followers = Dummy<Entity>.EmptyArray;
+
+        public Entity? Holding;
+
+        // TODO: Get rid of this, sync particles separately!
+        public (bool wasB, Vector2 dir)? Dash;
+
+        public bool Dead;
+
+        public override bool FilterHandle(DataContext ctx)
+            => Player != null; // Can be RECEIVED BY CLIENT TOO EARLY because UDP is UDP.
+
+        public override MetaType[] GenerateMeta(DataContext ctx)
+            => new MetaType[] {
+                new MetaPlayerUpdate(Player),
+                new MetaOrderedUpdate(Player?.ID ?? uint.MaxValue)
+            };
+
+        public override void FixupMeta(DataContext ctx) {
+            MetaPlayerUpdate playerUpd = Get<MetaPlayerUpdate>(ctx);
+            MetaOrderedUpdate order = Get<MetaOrderedUpdate>(ctx);
+
+            order.ID = playerUpd;
+            Player = playerUpd;
+        }
+
+        public override void ReadAll(CelesteNetBinaryReader reader) {
+            Player = reader.ReadRef<DataPlayerInfo>();
+
+            Flags flags = (Flags) reader.ReadByte();
+
+            Position = new(reader.Read7BitEncodedInt(), reader.Read7BitEncodedInt());
+            Scale = new(reader.ReadSByte() / 16f, reader.ReadSByte() / 16f);
+            Facing = ((flags & Flags.FacingLeft) != 0) ? Facings.Left : Facings.Right;
+            Speed = new(reader.ReadInt16(), reader.ReadInt16());
+
+            CurrentAnimationID = reader.Read7BitEncodedInt();
+            CurrentAnimationFrame = reader.Read7BitEncodedInt();
+
+            HairColors = new Color[reader.ReadByte()];
+            Color lastHairCol = Color.White;
+            for (int i = 0; i < HairColors.Length;) {
+                // Read the first color byte
+                byte firstColByte = reader.ReadByte();
+
+                // If it has the highest bit set, it's a repeat count of the last color
+                if ((firstColByte & (1<<7)) != 0) {
+                    int repCount = firstColByte & ~(1<<7);
+                    for (int j = 0; i < HairColors.Length && j < repCount; j++)
+                        HairColors[i++] = lastHairCol;
+                } else {
+                    // Read the second color byte and combine them
+                    int packedCol = (firstColByte << 8) | reader.ReadByte();
+
+                    // Unpack the color (each component is 5 bits)
+                    lastHairCol.R = (byte) (((packedCol >> 0) & 0b11111) << 3);
+                    lastHairCol.G = (byte) (((packedCol >> 5) & 0b11111) << 3);
+                    lastHairCol.B = (byte) (((packedCol >> 10) & 0b11111) << 3);
+                    lastHairCol.A = 255;
+
+                    HairColors[i++] = lastHairCol;
+                }
```

Comment:
I am not a fan of this at all, but if you really insist on squeezing out every last bit out of player frames... I won't be surprised if some day someone ends up complaining about alpha being lost though.

Reply:
I could add another 5-bit alpha component if you want